File: Amit-Kumar-Singh/Binary_Search/Find_First_And_Last_Position_Of_Element_In_Sorted_Array.java
//Time Compelxity: O(logn)
//Space Complexity: O(1)
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int mid = binarySearchNormal(nums,target);
        int targetStartIdx = -1;
        int targetEndIdx = nums.length;

        if(mid==-1){
            return new int[]{-1,-1};
        }
        targetStartIdx = binarySearchLeft(nums,0,mid-1,target);
        targetEndIdx = binarySearchRight(nums,mid+1,nums.length-1,target);
        return new int[]{targetStartIdx,targetEndIdx-1};
    }
    public static int binarySearchNormal(int[] nums, int target){
        int start = 0;
        int end = nums.length-1;
        while(start<=end){
            int mid = start + (end-start)/2;
            if(nums[mid]==target){
                return mid;
            }
            else if(nums[mid]>target){
                end = mid-1;
            }
            else{
                start = mid+1;
            }
        }
        return -1;
    }
    public static int binarySearchLeft(int[] nums, int start, int end, int target){
        while(start<=end){
            int mid = start+(end-start)/2;
            if(nums[mid]==target){
                end = mid-1;
            }
            else{
                start = mid+1;
            }
        }
        return start;
    }
    public static int binarySearchRight(int[] nums,int start,int end,int target){
        while(start<=end){
            int mid = start+(end-start)/2;
            if(nums[mid]==target){
                start = mid+1;
            }
            else{
                end = mid-1;
            }
        }
        return end+1;
    }
}

File: NikhilKumarGupta/Binary search/Find First and Last Position of Element in Sorted Array .cpp
class Solution {
    public:
    int first(vector<int>& nums, int target){
           int n=nums.size();
    
            int in1=-1;
            int s=0;
            int e=n-1;
           
           
            while(s<=e){
                int mid=(s+e)/2;
    
                if(target==nums[mid]){
                   in1=mid;
          e= mid - 1;
    
                    }
                else if(target>nums[mid]){
                    s=mid+1;
                }
                else{
                    e=mid-1;
                }
                
            }
            return in1;
    
    }
    int last(vector<int>& nums, int target){
           int n=nums.size();
    
            int in2=-1;
            int s=0;
            int e=n-1;
           
           
            while(s<=e){
                int mid=(s+e)/2;
    
                if(target==nums[mid]){
                   in2=mid;
          s= mid + 1;
    
                    }
                else if(target>nums[mid]){
                    s=mid+1;
                }
                else{
                    e=mid-1;
                }
                
            }
            return in2;
    
    }
        vector<int> searchRange(vector<int>& nums, int target) {
          int in1=first( nums, target);
          if(in1==-1){
            return {-1,-1};
    
          }
           int in2=last( nums, target);
           return {in1,in2};
    
           
        
        }
    };